### PR TITLE
Improve profanity filter

### PR DESCRIPTION
**Pull request by Kaltzisp**
**CONTEXT:**
Profanity filtering in BM has two components:
- Slur blocker (prevents you from sending text containing a slur)
- The actual profanity filter (replaces slurs & swears with asterisks ****)

**CHANGES:**
- Moved the slur blocker logic from server-side to client-side. This means we only need one `slurs.json` file, so I deleted the duplicate on the server-side.
- Refactored and improved the `profanityFilter` logic. Increased coverage and reduced false positives.
- Created a common directory at `react_main/src/lib` to hold shared scripts for the client.

image
image

Implements #711

### CODE REVIEW

**Comment by pikulet** [comment]
@monkney @milkshakesatdawn

File: react_main/src/lib/profanity.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            return segment;
+    }
+    // Substituting numbers with letters.
+    for (const num in leetMappings) {
+        segment = segment.replace(num, leetMappings[num]);
+    }
+    // Filtering profanity.
+    for (const profanityRegex of profanityRegexps) {
+        let regexRes = profanityRegex.exec(segment);
+        console.log(regexRes);
```
Comment (pikulet) [inline react_main/src/lib/profanity.js:71]: xx

**Review by DrSharky — DISMISSED** [review]

**Comment by milkshakesatdawn** [comment]
yeah, this implementation looks good to me, only thing I'd add is not to worry too much about chasing every single plausible character substitution down, there'll always be a way to bypass the filter for bad-faith users who are determined enough.

if and when that happens, that's fine to leave up to manual moderation to deal with IMO

File: react_main/src/lib/profanity.js (hunk 1 above)
Comment (Kaltzisp) [reply react_main/src/lib/profanity.js:71]: IDK WHAT XX MEANS T_T BUT IM GONNA REMOVE THE CONSOLE.LOG

File: react_main/src/lib/profanity.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        .map((word) => word.split("").join(String.raw`+([^A-Z]|\s)*`).concat("+"))
+        .map((word) => new RegExp(String.raw`\b[^A-Z]*((${word})+s*)+[^A-Z]*\b`, "i"));
+}
+
+// Creating profanity RegExps.
+const slurRegexps = createProfanityRegexps(slurs);
+const swearRegexps = createProfanityRegexps(swears);
+
+// Leet speak mappings.
+const leetMappings = {
```
[the remark this answers is not in the document]
Comment (Kaltzisp) [reply react_main/src/lib/profanity.js:28]: Honestly I'm happy either way, I don't think I've ever seen this substitution but apparently both 6 and 9 can be G. No real r reason not to e I'm going to add both.

**Comment by Kaltzisp** [comment]
> yeah, this implementation looks good to me, only thing I'd add is not to worry too much about chasing every single plausible character substitution down, there'll always be a way to bypass the filter for bad-faith users who are determined enough.
> 
> if and when that happens, that's fine to leave up to manual moderation to deal with IMO

Agreed - could add symbol mappings (@==A, !==I) but I think that's starting to be overkill, as you'd need to check it both against the original text and also against the substituted (so that "slur!!!" doesn't get interpreted as "sluriii".

There's always going to be a way around the restriction, but I'm sufficiently happy that this should at least make anyone that tries to use a slur and is blocked from doing so on the first try, and has time to think twice before commenting.

**Comment by Kaltzisp** [comment]
Also thanks for reviewing @DrSharky @pikulet , made some changes so sorry to have to ask you to come back!

**Review by pikulet — APPROVED** [review]

**Pull request merged**